### ISSUE TITLE
Add option for return key adding line break, and double enter adding new paragraph

### ISSUE THREAD

**Issue by sathishmanohar** [yabwe/medium-editor#70]
Currently pressing the return key once creates new paragraph, which I feel is counter intuitive. By defaulting to line breaks on single return and paragraphs on double returns, It'll feel lot more intuitive, since many other editors and word processors use the same pattern.

**Comment by bbnnt** [comment]
@litera is shift+enter is adding a br tag when using medium editor ?

**Comment by hellocosmin** [comment]
@bbnnt yes it does, you can check the demo using a browser inspection tool.

Since someone brought this up, I have another (related) concern. Here's the scenario:

- I have a .editable on which I use ME; this element is a block level (h1, div...)
- using <kbd>Shift+Enter</kbd> correctly adds a `<br>`
- hitting <kbd>Backspace</kbd> to undo/remove that `<br/>` wraps the entire .editable contents inside a `<p></p>`

This results in incorrect formatting, since I now have a paragraph inside my h1, for example. Might be a smaller issue for web view, but in email design this is a nightmare.
I know there are ways of dealing with this (like unwrapping the contents or using something like jQuery htmlclean() plugin ), but could we consider a possible option for ME, by which we can turn paragraph wrapping on/off?

ME can be used in so many applications and it would be a really useful option to have (especially in email layout editors), instead of having to manually clean HTML or removing some code in the core.

**Comment by litera** [comment]
@bbnnt As @hellocosmin pointed out, the answer to your question is **yes, <kbd>Shift + ENTER</kbd> creates a `<br />`**.

@hellocosmin: This issue you're having is very likely due to editable behaviour of specific browser. You can check individual browsers if they all act the same. but in order to solve your issue you can always *unwrap* your content if wrapping is inside a `p` element.

I've had an issue with Chrome browser creating `span` elements with inline styles that was particularly annoying because it was setting `font-size` style. So I added my code that unwrapped `span` content and everything's ok now. You can do the same for your paragraph wrapping.

**Comment by hellocosmin** [comment]
@litera Checked in both Chrome and Firefox, it does exactly the same thing.

Commenting out this line, makes it work fine (i.e. not wrapping the contents in a paragraph):
https://github.com/daviferreira/medium-editor/blob/master/dist/js/medium-editor.js#L1799

**Comment by bbnnt** [comment]
@litera @hellocosmin  thanks a lot lot for your hints
unfortunately, the shift+enter does creates a br :/
hope I'll get over this
I'll check this span thing that occurs on my side that is extremely (!!) annoying

**Comment by bbnnt** [comment]
@hellocosmin it does not on safari through. no line breaks possibles.

**Comment by nmielnik** [comment]
To clarify this issue, I believe the correct way to summarize this would be:

Expose option to have pressing <kbd>ENTER</kbd> insert a `<br>` tag, instead of creating a new `<p>` by default.  

Something like a `singleEnterBlockElement` option, which is `true` by default, but when turned off, will insert a `<br>` for single <kbd>ENTER</kbd>, and a `<p>` for <kbd>ENTER</kbd> x 2

**Comment by frankrousseau** [comment]
Hello there,

  Thank you for Medium-editor, it's really a nice and beautiful editor. So good, that our team integrated to the [webmail](https://github.com/cozy/cozy-emails) we are building. Because it's a webmail we would like to have the behavior of `shift+enter` when only `enter` is typed. I wanted to modify directly the source code of the version we embed, but I can't find what should be changed. 

Commenting https://github.com/yabwe/medium-editor/blob/master/src/js/core.js#L142 and https://github.com/yabwe/medium-editor/blob/master/src/js/core.js#L153 doesn't change anything. 

Do you have any idea of what should be changed for return key adding line break?

**Comment by nmielnik** [comment]
So https://github.com/yabwe/medium-editor/blob/master/src/js/core.js#L14 is closer along the lines of what you would want to do.  This is where the code handles preventing newlines when <kbd>ENTER</kbd> is pressed (`disableEnter` and `disableDoubleEnter`).

Your custom code could manually subscribe to `editableKeydownEnter` like we do [here](https://github.com/yabwe/medium-editor/blob/master/src/js/core.js#L360) which is where we detect if `disableEnter` or `disableDoubleEnter` options have been added.

Once you've detected <kbd>SHIFT</kbd> + <kbd>ENTER</kbd>, you'll want to call `preventDefault()` and then add your logic for adding a linebreak the way you want.

If you get this working, this is definitely something that we would love to have in MediumEditor.  As you can see by this issue, there have definitely been requests for it.  We would just want a new Option so that you could toggle it on and off.

Hope that helps.

**Comment by frankrousseau** [comment]
Thank you for your detailed answer. I already tried to comment these lines and add a `insertHTML` command with `<br>` inside. It works but it messes up everything with the focus and the br is not placed wher I would like. 
What I noticed is that when you do `shift + enter`, it's the expected behavior. So my question is how can i bind this behavior directly to the `enter` command?

**Comment by nmielnik** [comment]
Outside of handling keydown on <kbd>ENTER</kbd> and inserting the HTML yourself, I'm not sure of a way to do this.  I don't believe you can mimic the browser's built-in handling of <kbd>SHIFT</kbd> + <kbd>ENTER</kbd> but it could be worth doing a little searching to see how others may have handled it.

**Comment by frankrousseau** [comment]
Ok it's the browser which performs the change not the editor. I understand now why I didn't find the code for it. So, I will go that way. Thank you.

**Comment by nmielnik** [comment]
This has also been talked about in other issues, such as #34 and #251

**Comment by sedletsky** [comment]
I'm trying to implement Medium editor behavior: I need to insert line break (<br>) on Shift+Enter and clone paragraph (<p>) on Enter. I have editable section with paragraphs. I wrote these code:

var e = new MediumEditor("[contenteditable=true]")
e.subscribe("editableKeydownEnter", function (event, element) {
    if (event.shiftKey) {
        event.preventDefault()
        var node = MediumEditor.selection.getSelectionStart(this.options.ownerDocument)
        MediumEditor.util.insertHTMLCommand(this.options.ownerDocument, "<br>")
    }
}.bind(e))

It doesn't work properly. Where is mistake? I losing a hope.

**Comment by nmielnik** [comment]
@sedletsky can you verify that this wasn't the default behavior that browsers implemented?  

As far as I've seen and heard, when you hit <kbd>ENTER</kbd> inside a `contenteditable` element, the browsers inserted new `<p>` elements, and when you hit <kbd>SHIFT</kbd> + <kbd>ENTER</kbd> it would insert a `<br/>`.  Based on that, it doesn't seem like you would need to add any functionality.

**Comment by lordmaul** [comment]
@sedletsky Try the following. It should work
editor.subscribe("editableKeydownEnter", function (event, element) {
				if (event.shiftKey) {
					
					var node = MediumEditor.selection.getSelectionStart(editor.options.ownerDocument)
					MediumEditor.util.insertHTMLCommand(editor.options.ownerDocument, "gello")
					event.preventDefault()
				}
			}.bind(e))

**Comment by timurguseynov** [comment]
This might work:

`var e = new MediumEditor("[contenteditable=true]")
e.subscribe("editableKeydownEnter", (event, element) => {
    if (event.shiftKey) {
        event.preventDefault()
        event.stopPropagation()
        var node = MediumEditor.selection.getSelectionStart(this.options.ownerDocument)
        MediumEditor.util.insertHTMLCommand(this.options.ownerDocument, "<br>")
    }
}`

**Comment by yepecece** [comment]
> This might work:
> 
> ```js
> e.subscribe("editableKeydownEnter", (event, element) => {
>     if (event.shiftKey) {
>         event.preventDefault()
>         event.stopPropagation()
>         MediumEditor.selection.getSelectionStart(this.options.ownerDocument)
>         MediumEditor.util.insertHTMLCommand(this.options.ownerDocument, "<br>")
>     }
> }
> ```

This correctly adds a `<br>` but it closes the `<p>` tag before inserting it (in safari).
I'm trying to fix the soft return bug in safari with this. #1119 

Any idea as to how to keep the `<p>` tag after the `<br>` on soft return?

**Comment by DavidGeismarLtd** [comment]
This

> > This might work:
> > ```js
> > e.subscribe("editableKeydownEnter", (event, element) => {
> >     if (event.shiftKey) {
> >         event.preventDefault()
> >         event.stopPropagation()
> >         MediumEditor.selection.getSelectionStart(this.options.ownerDocument)
> >         MediumEditor.util.insertHTMLCommand(this.options.ownerDocument, "<br>")
> >     }
> > }
> > ```
> 
> This correctly adds a `<br>` but it closes the `<p>` tag before inserting it (in safari). I'm trying to fix the soft return bug in safari with this. #1119
> 
> Any idea as to how to keep the `<p>` tag after the `<br>` on soft return?

Also this adds a `<br>` but doesnt move the user cursor after the `br`, any idea how to do that ?
